src/util/invariant.cpp

```diff
@@ -0,0 +1,152 @@
+/*******************************************************************\
+
+Module:
+
+Author: Martin Brain, martin.brain@diffblue.com
+
+\*******************************************************************/
+
+
+#include "invariant.h"
+
+#include <stdexcept>
+#include <string>
+#include <sstream>
+
+
+#ifdef CPROVER_INVARIANT_PRINT_STACK_TRACE
+#include <iostream>
+#include <stdlib.h>
+#endif
+
+// Backtraces compiler and C library specific
+#include <assert.h>
+#ifdef __GLIBC__
+
+// GCC needs LINKFLAGS="-rdynamic" to give function names in the backtrace
+#include <execinfo.h>
+#include <cxxabi.h>
+
+
+/*******************************************************************\
+
+Function: output_demangled_name
+
+  Inputs: The output stream and the description of the stack_entry
+
+ Outputs: True <=> the entry has been successfully demangled and printed.
+
+ Purpose: Attempts to demangle the function name assuming the glibc
+          format of stack entry:
+
+          path '(' mangled_name '+' offset ') [' address ']\0'
+
+\*******************************************************************/
+
+static bool output_demangled_name(std::ostream &out,
+                                  const char * const stack_entry)
+{
+  bool return_value = false;
+
+  std::string working(stack_entry);
+
+  size_t start = working.rfind('(');  // Path may contain '(' !
+  size_t end = working.find('+', start);
+
+  if(start != std::string::npos &&
+     end != std::string::npos &&
+     start + 1 <= end - 1)
+  {
+    size_t length = end - (start + 1);
+    std::string mangled(working.substr(start + 1, length));
+
+    int demangle_success = 1;
+    char *demangled =
```

Comment:
I'd rather keep these as free as the documentation says that they are explicitly malloc'd.